storage/store.go

```diff
@@ -927,27 +930,38 @@ func (s *Store) NewRangeDescriptor(start, end proto.Key, replicas []proto.Replic
 
 // SplitRange shortens the original range to accommodate the new
 // range. The new range is added to the ranges map and the rangesByKey
-// sorted slice.
+// btree.
 func (s *Store) SplitRange(origRng, newRng *Range) error {
 	if !bytes.Equal(origRng.Desc().EndKey, newRng.Desc().EndKey) ||
 		bytes.Compare(origRng.Desc().StartKey, newRng.Desc().StartKey) >= 0 {
 		return util.Errorf("orig range is not splittable by new range: %+v, %+v", origRng.Desc(), newRng.Desc())
 	}
+
+	s.mu.Lock()
 	// Replace the end key of the original range with the start key of
-	// the new range.
+	// the new range. Reinsert the range since the btree is keyed by range end keys.
+	if s.rangesByKey.Delete((*rangeBTreeItem)(origRng)) == nil {
+		s.mu.Unlock()
```

Comment:
`processRangeDescriptorUpdate` acquires the lock, so we need to release the lock before calling the method..